### PR TITLE
Extract all expect interface methods into common

### PR DESCRIPTION
**Pull request by Egorand**

### CODE REVIEW

**Comment by rjrjr** [comment]
Did you mean to be merging this into `egorand/190404/expect-buffer-timeout` instead of master?

**Comment by Egorand** [comment]
Yes!

File: okio/src/main/kotlin/okio/Source.kt [hunk 1]
```diff
@@ -15,4 +15,8 @@
  */
 package okio
 
-expect interface Source
+expect interface Source {
+  fun read(sink: Buffer, byteCount: Long): Long
+
```
Comment (JakeWharton) [inline okio/src/main/kotlin/okio/Source.kt]: Missing close

File: okio/src/main/kotlin/okio/Sink.kt [hunk 2]
```diff
@@ -15,4 +15,8 @@
  */
 package okio
 
-expect interface Sink
+expect interface Sink {
+  fun write(source: Buffer, byteCount: Long)
+
```
Comment (JakeWharton) [inline okio/src/main/kotlin/okio/Sink.kt]: Missing close and flush

File: okio/src/main/kotlin/okio/BufferedSource.kt [hunk 3]
```diff
@@ -15,4 +15,82 @@
...
+
+  fun readDecimalLong(): Long
+
+  fun readHexadecimalUnsignedLong(): Long
+
+  fun skip(byteCount: Long)
+
+  fun readByteString(): ByteString
+
+  fun readByteString(byteCount: Long): ByteString
```
Comment (JakeWharton) [inline okio/src/main/kotlin/okio/BufferedSource.kt]: Lots of cases where default argument values with `@JvmOverloads` should be used here in common. Also in BufferedSink.

File: okio/src/main/kotlin/okio/Buffer.kt [hunk 4]
```diff
@@ -15,4 +15,131 @@
...
+
+  override fun readIntLe(): Int
+
+  override fun readLongLe(): Long
+
+  override fun readDecimalLong(): Long
+
+  override fun readHexadecimalUnsignedLong(): Long
+
+  override fun readByteString(): ByteString
```
Comment (JakeWharton) [inline okio/src/main/kotlin/okio/Buffer.kt]: None of these functions should need redeclare unless they're using covariant return types to change the return type to `Buffer`

Comment (Egorand) [reply okio/src/main/kotlin/okio/Buffer.kt]: Oh, good call. I was getting build failures so pulled everything out without digging deep into specific methods. 

Covariant returns are a bit inconsistent due to the fact that `Buffer` omits return types in many places, so it's hard to say whether using covariant returns was intentional or not.

File: okio/src/main/kotlin/okio/BufferedSource.kt (hunk 3 above)
Comment (Egorand) [reply okio/src/main/kotlin/okio/BufferedSource.kt]: Can address in more targeted PRs

File: okio/src/main/kotlin/okio/Sink.kt (hunk 2 above)
Comment (Egorand) [reply okio/src/main/kotlin/okio/Sink.kt]: Those come from Java interfaces so keeping them in JVM for now. Can also introduce our own interfaces and typealias to Java ones, would that make sense?

Comment (JakeWharton) [reply okio/src/main/kotlin/okio/Sink.kt]: On the JVM we use the interfaces to carry forward the declarations, but those methods are still intrinsic to the type and need defined here. What is a sink or a source that you can't call close on?

File: okio/src/main/kotlin/okio/BufferedSource.kt (hunk 3 above)
Comment (JakeWharton) [reply okio/src/main/kotlin/okio/BufferedSource.kt]: File an issue so we don't lose track? I wouldn't want to ship a version like this

File: okio/src/main/kotlin/okio/Buffer.kt (hunk 4 above)
Comment (JakeWharton) [reply okio/src/main/kotlin/okio/Buffer.kt]: I think it should just be for all the `write` methods to support chaining on the JVM without running into the checked exceptions from `BufferedSink`

File: okio/src/main/kotlin/okio/Buffer.kt [hunk 5]
```diff
@@ -15,4 +15,135 @@
  */
 package okio
 
-expect class Buffer : BufferedSource, BufferedSink
+expect class Buffer : BufferedSource, BufferedSink {
```
Comment (rjrjr) [inline okio/src/main/kotlin/okio/Buffer.kt]: Pretty sure you want `Buffer()` here, the zero args constructor is public.

File: okio/js/src/main/kotlin/okio/Buffer.kt [hunk 6]
```diff
@@ -15,4 +15,149 @@
  */
 package okio
 
-actual class Buffer : BufferedSource, BufferedSink
+actual class Buffer : BufferedSource, BufferedSink {
```
Comment (swankjesse) [inline okio/js/src/main/kotlin/okio/Buffer.kt]: FWIW I think Buffer fits better in common completely.

File: okio/js/src/main/kotlin/okio/Sink.kt [hunk 7]
```diff
@@ -15,4 +15,12 @@
  */
 package okio
 
-actual interface Sink
+actual interface Sink {
```
Comment (swankjesse) [inline okio/js/src/main/kotlin/okio/Sink.kt]: can we just put these interfaces in common?

File: okio/src/main/kotlin/okio/Source.kt (hunk 1 above)
Comment (swankjesse) [reply okio/src/main/kotlin/okio/Source.kt]: Fixed.

File: okio/js/src/main/kotlin/okio/Sink.kt (hunk 7 above)
Comment (rjrjr) [reply okio/js/src/main/kotlin/okio/Sink.kt]: I don't think so, b/c of the java.io deps.

File: okio/js/src/main/kotlin/okio/Buffer.kt (hunk 6 above)
Comment (rjrjr) [reply okio/js/src/main/kotlin/okio/Buffer.kt]: Again, lots of java deps. Unless it's possible to have a non-`expects` common class enhanced by a platform specific one? Didn't think that was the case.

Comment (Egorand) [reply okio/js/src/main/kotlin/okio/Buffer.kt]: That would be the use case for concrete implementations being allowed in `expect` classes, which is not there at the moment. For now we'll have to work around it in the same way we did with `ByteString`.

File: okio/js/src/main/kotlin/okio/Sink.kt (hunk 7 above)
Comment (Egorand) [reply okio/js/src/main/kotlin/okio/Sink.kt]: Right, we're adding supertypes for `Sink` on the JVM side and need an `actual` for this.

**Pull request merged**